### ISSUE TITLE
Blank values (not zero) on older pandas

### ISSUE THREAD

**Issue by peterjc** [abaizan/kodoja#29]
Following 2646bd59d01d681887e9ce6b65d44e7cf31a8b47 which I intend to include in kodoja v0.0.7, we expect zeros rather thank blanks in columns 6 and 7 of ``virus_table.txt`` from ``kodoja_search.py``

That change seems to work fine on pandas 0.20 onwards,

```
$ pip3 install pandas==0.20 && python3 diagnosticTool_scripts/kodoja_search.py -r1 test/data/testData_1.fastq -o test/PE_test/ -d1 test/example_db/krakenDB_test/ -d2 test/example_db/kaijuDB_test/ -r2 ./test/data/testData_2.fastq -t 2 && diff test/PE_test/virus_table.txt test/data/virus_table_PE_fastq.txt && diff test/PE_test/kodoja_VRL.txt test/data/virus_reads_PE_fastq.txt
...
Started analysis of trimmed_read1
Analysis complete for trimmed_read1
Started analysis of trimmed_read2
Analysis complete for trimmed_read2
94 sequences (0.02 Mbp) processed in 0.021s (269.9 Kseq/m, 54.62 Mbp/m).
  94 sequences classified (100.00%)
  0 sequences unclassified (0.00%)
16:32:58 Reading database
 Reading taxonomic tree from file test/example_db/krakenDB_test/taxonomy/nodes.dmp
 Reading index from file test/example_db/kaijuDB_test/kaiju_library.fmi
Output file: test/PE_test/kaiju_table.txt
16:32:58 Start classification using 2 threads.
16:32:58 Finished.
```

But it fails on pandas 0.19 and older (note using 0.16 or older requires a tweak to the column sorting),

```
$ pip3 install pandas==0.19 && python3 diagnosticTool_scripts/kodoja_search.py -r1 test/data/testData_1.fastq -o test/PE_test/ -d1 test/example_db/krakenDB_test/ -d2 test/example_db/kaijuDB_test/ -r2 ./test/data/testData_2.fastq -t 2 && diff test/PE_test/virus_table.txt test/data/virus_table_PE_fastq.txt && diff test/PE_test/kodoja_VRL.txt test/data/virus_reads_PE_fastq.txt
...
Started analysis of trimmed_read1
Analysis complete for trimmed_read1
Started analysis of trimmed_read2
Analysis complete for trimmed_read2
94 sequences (0.02 Mbp) processed in 0.022s (262.2 Kseq/m, 53.08 Mbp/m).
  94 sequences classified (100.00%)
  0 sequences unclassified (0.00%)
16:36:04 Reading database
 Reading taxonomic tree from file test/example_db/krakenDB_test/taxonomy/nodes.dmp
 Reading index from file test/example_db/kaijuDB_test/kaiju_library.fmi
Output file: test/PE_test/kaiju_table.txt
16:36:04 Start classification using 2 threads.
16:36:04 Finished.
2,4c2,4
< Cassava brown streak virus	137758	45	45	Ipomovirus		
< Ugandan cassava brown streak virus	946046	28	28	Ipomovirus		
< Tobacco etch virus	12227	21	19	Potyvirus		
---
> Cassava brown streak virus	137758	45	45	Ipomovirus	0	0
> Ugandan cassava brown streak virus	946046	28	28	Ipomovirus	0	0
> Tobacco etch virus	12227	21	19	Potyvirus	0	0
```

**Comment by peterjc** [comment]
Reduced test case,

```
$ pip3 install pandas==0.20 && python3 -c "from collections import Counter; import pandas as pd; df=pd.DataFrame([[1,123],[1,1],[4,2],[5,22]], columns=['a','b']); d = Counter(); d[1]=111; d[22]=2; print(df['b'].map(d)); print([d[_] for _ in df['b']])"
Collecting pandas==0.20
Requirement already satisfied: python-dateutil>=2 in /mnt/shared/users/pc40583/lib/python3.5/site-packages (from pandas==0.20) (2.6.0)
Requirement already satisfied: pytz>=2011k in /mnt/shared/users/pc40583/lib/python3.5/site-packages (from pandas==0.20) (2017.2)
Requirement already satisfied: numpy>=1.7.0 in /mnt/shared/users/pc40583/lib/python3.5/site-packages (from pandas==0.20) (1.14.2)
Requirement already satisfied: six>=1.5 in /mnt/shared/users/pc40583/lib/python3.5/site-packages (from python-dateutil>=2->pandas==0.20) (1.10.0)
Installing collected packages: pandas
  Found existing installation: pandas 0.19.0
    Uninstalling pandas-0.19.0:
      Successfully uninstalled pandas-0.19.0
Successfully installed pandas-0.20.0
0      0
1    111
2      0
3      2
Name: b, dtype: int64
[0, 111, 0, 2]
```

Note that the second column above is 0, 111, 0, 2 matching the list output, but on the older pandas this fails and we get NaN entries:

```
$ pip3 install pandas==0.19 && python3 -c "from collections import Counter; import pandas as pd; df=pd.DataFrame([[1,123],[1,1],[4,2],[5,22]], columns=['a','b']); d = Counter(); d[1]=111; d[22]=2; print(df['b'].map(d)); print([d[_] for _ in df['b']])"
Collecting pandas==0.19
  Using cached https://files.pythonhosted.org/packages/dc/05/890be6bd59f39952291f3e50d8dbd0adabda27e01c15dfcf3968f6902a44/pandas-0.19.0-cp35-cp35m-manylinux1_x86_64.whl
Requirement already satisfied: numpy>=1.7.0 in /mnt/shared/users/pc40583/lib/python3.5/site-packages (from pandas==0.19) (1.14.2)
Requirement already satisfied: pytz>=2011k in /mnt/shared/users/pc40583/lib/python3.5/site-packages (from pandas==0.19) (2017.2)
Requirement already satisfied: python-dateutil>=2 in /mnt/shared/users/pc40583/lib/python3.5/site-packages (from pandas==0.19) (2.6.0)
Requirement already satisfied: six>=1.5 in /mnt/shared/users/pc40583/lib/python3.5/site-packages (from python-dateutil>=2->pandas==0.19) (1.10.0)
Installing collected packages: pandas
  Found existing installation: pandas 0.20.0
    Uninstalling pandas-0.20.0:
      Successfully uninstalled pandas-0.20.0
Successfully installed pandas-0.19.0
0      NaN
1    111.0
2      NaN
3      2.0
Name: b, dtype: float64
[0, 111, 0, 2]

**Comment by peterjc** [comment]
https://pandas.pydata.org/pandas-docs/version/0.19.0/generated/pandas.Series.map.html

vs

https://pandas.pydata.org/pandas-docs/version/0.20.3/generated/pandas.Series.map.html

added:

> When arg is a dictionary, values in Series that are not in the dictionary (as keys) are converted to NaN. However, if the dictionary is a dict subclass that defines __missing__ (i.e. provides a method for default values), then this default is used rather than NaN:

This is the behaviour I was relying on, noted in the whats-new page https://pandas.pydata.org/pandas-docs/version/0.20.3/whatsnew.html referencing their github issue 15999 which noted a workaround using a lambda function instead.

I will explore using the lambda trick on my branch getting kodoja to work with older versions of pandas...

**Comment by peterjc** [comment]
Fixed in 3683042207be880c68fbcf260f9a6040e44a2695
